lectures/1-Introduction/LinearAlgebra.md

```diff
@@ -0,0 +1,240 @@
+(\lambda v) \otimes w = v \otimes (\lambda w) = \lambda (v \otimes w)
+```
+
+```{code-cell} julia
+λ = rand()
+(λ * v) ⊗ w ≈ v ⊗ (λ * w) ≈ λ * (v ⊗ w)
+```
+
+This new vector space can be equipped with a canonical basis, which is constructed by taking
+the tensor product of the basis vectors of the original vector spaces. For example, if $V$
+and $W$ are two-dimensional vector spaces with basis vectors $v_i$ and $w_j$, respectively,
+then the basis vectors of $V \otimes W$ are given by $v_i \otimes w_j$. In other words, the
+vectors in $V \otimes W$ are linear combinations of all combinatinos of the basis vectors of
+$V$ and $W$.
+
+When considering how to represent a vector in this new vector space, it can be written as a
+list of numbers that correspond to the components of the vector in that basis. For example,
+a vector in $V \otimes W$ is described by:
+
+```{math}
+:label: eq:tensor_basis
+t = \sum_{i_1,i_2} t_{i_1i_2} (v_{i_1} \otimes w_{i_2})
+```
+
+```{code-cell} julia
+t = Tensor(rand, Float64, V ⊗ W)
+t[] # shorthand for extracting the multi-dimensional array of components
+```
+
+Here, the tentative name $t$ was used to denote that this is in fact a tensor, where
+$t_{i_1i_2}$ are the components of that tensor $t$ in the basis $v_{i_1} \otimes w_{i_2}$.
+Because of the induced structure of the tensor product, it is more natural and very common
+to express this object not just as a list of numbers, but by reshaping that list into a
+matrix. In this case, the components of the $i_1$-th row correspond to basis vectors that
+are built from $v_{i_1}$, and similarly the $i_2$-th column corresponds to basis vectors
+that are built from $w_{i_2}$.
+
+As the tensor product can be generalized to more than two vector spaces, this finally leads
+to the general definition of a tensor as an element of the vector space that is built up
+from the tensor product of an arbitrary number of vector spaces. Additionally, the
+components of these objects are then naturally laid out in a multi-dimensional array, which
+is then by a slight misuse of terminology also called a tensor.
+
+```{note}
+The reshaping operation of components from a list of numbers into a multi-dimensional array
+is nothing but a mapping between linear indices $I$ and Cartesian indices $i_1, i_2, \cdots,
+i_N$. This is a very common and useful trick which allows reinterpreting tensors as vectors,
+or vice versa.
+```
+
+```{code-cell} julia
+LinearIndices((1:2, 1:3))
+```
+
+```{code-cell} julia
+collect(CartesianIndices((1:2, 1:3))) # collect to force printing
+```
+
+## Tensors and Multi-Linear Maps
+
+Due to the fact that the tensor product of vector spaces is a vector space in of itself, it
+is again possible to define linear maps between such vector spaces. Keeping in mind the
+definition of a linear map from {eq}`eq:linear_map`, the columns now label components of the
+input vector space, while the rows label components of the output vector space. Now however,
```

Comment:
Introduce terminology of 'domain' and 'codomain' for input and output vector spaces respectively?